demo style tweak: Extend inpage footnotes

moben opened the pull request:
Draft / Demo for https://github.com/koreader/crengine/pull/618 to allow extending footnotes across lines from consecutive nodes. For now this is just a reference of how that could be used in koreader

Ideally I'd like to land the `-cr-only-if: following-inpage-footnote` version if that's accepted in crengine. The other version is merely an illustration of how a subset of that can also be achieved via css selectors.

I'll update this after discussion in crengine.

Suggestions welcome where it should live, thought that probably depends on which version of this would be accepted. The second version would have to be per footnote type, the first could be a generic _"footnote fix-up"_.

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1031,6 +1031,46 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+            css = [[
+*, autoBoxing {
+    -cr-hint: late;
+    -cr-only-if: following-inpage-footnote -inpage-footnote;
+    -cr-hint: extend-footnote-inpage;
+    margin: 0 !important;
```
More indentation on the lines following a `-cr-only-if:` (to make it visually clear the next apply only-if :) ala python.
cf the existing ones.

poire-z commented:
If you add 2 items, we'll get 11 items, with one on the 2nd page (in landscape mode).
If you really need these 2 new ones, it feels best to create a submenu (so, 3 submenus at bottom), ie. "In-page footnote extension".

moben commented:
I'm fine with a submenu if you're willing to have multiple versions of this tweak. (and maybe someone comes up with another one in the future)

I do think having the two versions  "until next footnote" (aka `*, autoBoxing`) and "until next heading or footnote" (aka `*:not(h1,...), autoBoxing` makes sense because the former is the maximalist version that makes sure you never miss any footnotes and the latter is close to what popup footnotes do so there's some precedent for this being a good stopping point (and I have at least one actual book where it's the correct choice).

Having a version with `:not()` as an example to copy from might also be nice for people who aren't css experts to have an easier time writing user / book tweaks.

moben commented:
1. Moved footnote extension to submenu
2. Fixed "smaller" variants of in-page footnotes in the presence of extended footnotes.

I think the fix for (2) is somewhat dependent on order / selector specificity but seems to work at least in my testing. I think the only possible alternative (that doesn't need a new crengine feature) is adding "smaller" variants for each footnote extension and the user having to make sure they match if they want consistent styling.

moben commented:
Marked as non-draft on the basis that I think https://github.com/koreader/crengine/pull/618 is pretty much ready.

moben commented:
@Commodore64user replying to you here instead of https://github.com/koreader/koreader/pull/13613#issuecomment-2812358410 to keep the discussion more focused.

> [...], although the `extension` stuff could go inside `fix-up` easily. Just add a separator and whack 'em at the bottom.

@poire-z and myself discussed that here. The reason I put them outside fix-ups is that fix-ups are all about how footnotes are styled, whereas the top level settings (and the new extension stuff) are about which lines are selected to be footnotes. And there might be more common cases like "until the next heading" that could go in the same menu (not sure which, would need more people to use it and report back).

> I do find the names `extend footnotes until next footnote` and `extend footnotes until next header or footnote` a bit too overwhelming. I would suggest something more like `extend footnote selection/content up to next footnote` and `extend footnote selection up to next header`.

I tried that just now but I think it's too long. Everywhere I was able to test it's cut of and ellipsized.

moben commented:
Well, my head didn't read those as a choice so it was a bit long. (and maybe I made a copy&paste mistake) :sweat_smile: 
Obviously fits with your wording and is a bit clearer, happy to switch to that.

Maybe the "until the next footnote" part should just be in the description, so it can be

- `extend footnote content`
- `extend footnote content until next header`

Every entry in this submenu would be `until X or the next footnote` anyway.

Commodore64user commented:
better to have the first as `extend footnote content until next entry` the two should be somewhat similar in char length ;)

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -943,6 +943,15 @@ This only works with footnotes that have specific attributes set by the publishe
     -cr-only-if: -fb2-document;
         -cr-hint: footnote-inpage;
         margin: 0 !important;
+}
+*, autoBoxing {
+    -cr-hint: late;
+    -cr-only-if: inpage-footnote;
+        font-size: 0.8rem !important;
+}
+*, autoBoxing {
+    -cr-hint: late;
+    -cr-only-if: extended-inpage-footnote;
```
Can't the first `*, autoBoxing {` content still be held in the `*[role~="doc-rearnote"] {`, to avoid a CSS check ?

moben replied:
Sure. I think I was splitting it out because I was already thinking about https://github.com/koreader/koreader/pull/13613 and it needs to be separate there.

moben replied:
Reduced this to one extra CSS check with https://github.com/koreader/crengine/pull/620

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1026,11 +1044,54 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+                title = _("Extend footnote content until next entry"),
+                description = _([[
+Extend in-page footnotes shown at the bottom of pages to include text up to the next footnote.
+This is needed when books don't correctly mark all text that belongs to the footnote.]]),
```
may be / might be (because who are we to force inpage footnotes to the use ? :)) I think that's the tone we have in other tweak descriptions.

moben replied:
done

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1026,11 +1044,54 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+        {
+            title = _("In-page footnote extension"),
+            {
+                id = "following-footnote-inpage_all",
```
(Dunno if `_all` is standardized elsewhere, but `_any` -or nothing- feels more accurate?)

moben replied:
changed to `_any` (and adjusted the id prefix to `extend-footnote-inpage_following-inpage-footnote_` which I think fits with the existing naming scheme based on the `-cr-hint` / `-cr-only-if` that is applied or used.

poire-z replied:
Feels a bit long though :) (related to all the other ids).
`extend-footnote-inpage_following-inpage-footnote_ `
May be just ("following inpage footnote" seems implied with "extend footnote inpage"):
`extend-footnote-inpage_any`
`extend-footnote-inpage_until_next_heading`
?
(Not enough energy to review the rest this evening.)

poire-z replied:
Looks ok, but still waiting for these to get shorter :)

moben replied:
sorry, had forgotten about this. They are shorter now :+1:

poire-z commented on frontend/ui/data/css_tweaks.lua:
```diff
@@ -1026,11 +1044,54 @@ This tweak can be duplicated as a user style tweak when books contain footnotes
+Extend in-page footnotes shown at the bottom of pages to include text up to the next footnote.
+This is needed when books don't correctly mark all text that belongs to the footnote.]]),
+                css = [[
+*, autoBoxing {
+    -cr-only-if: following-inpage-footnote -inpage-footnote;
+        -cr-hint: extend-footnote-inpage;
```
It feels that this, having a really low specificity, may not match those that get to be inpage-footnotes with `*[type~="note"]` or `.footnote` which has a higher specificity.
Is `-cr-hint: late` the solution ? Dunno how everything behave when multiple cr-hint late setting other hints... moreover when one would use `-cr-only-if: extended-inpage-footnote`...

moben replied:
It's certainly more fragile, I'll do some testing with `-cr-hint: late`.

`-cr-only-if: extended-inpage-footnote` is essentially only needed to set the smaller font-size above for well behaved books where you don't need to force a single size via the `-cr-only-if: inside-inpage-footnote` hint.

We could also go back and make `-cr-only-if: inpage-footnote` also match extended footnotes. Less css but also less flexible. I at least found it useful in debugging to match the extended blocks specifically.

poire-z replied:
Logically, `-cr-only-if: inpage-footnote` (named `inpage-foonote`, different from the hint name `footnote-inpage`) should match what has ended up being a footnote, not what had the hint.
So yes, I think it should also match extended footnotes.

Same for `inside-inpage-footnote`, which should match stuff inside an extended footnote block.

(I'm out for the day.)

moben replied:
I agree, changed it here and dropped the unneeded `-cr-only-if: extended-inpage-footnote`: https://github.com/koreader/crengine/pull/620

moben replied:
I remembered why I didn't originally add `late`: I had problems with specificity like you hinted at.

I did what I think is the correct fix now by using `-cr-hint: late` to ensure it's applied after footnotes are marked as well as `:where()` and `priority` to ensure it's lower specificity than other `late` hints that style footnotes. (I only learned today that tweak order depends on the name of the tweak, so `priority` is actually necessary).

poire-z commented:
Will need a crengine into base bump, and a base into frontend bump.

Commodore64user commented:
![image](https://github.com/user-attachments/assets/83e4b4e9-2964-421a-b636-1a6166c68fb3)
